lib/dyno.js

```diff
@@ -62,28 +66,102 @@ class Dyno {
    */
   attach () {
     return new Promise((resolve, reject) => {
-      if (this.opts.showStatus) cli.action.status(this._status('starting'))
       this.resolve = resolve
       this.reject = reject
-      let uri = url.parse(this.dyno.attach_url)
-      let c = tls.connect(uri.port, uri.hostname, {rejectUnauthorized: this.heroku.options.rejectUnauthorized})
-      c.setTimeout(1000 * 60 * 20)
-      c.setEncoding('utf8')
-      c.on('connect', () => {
-        c.write(uri.path.substr(1) + '\r\n', () => {
-          if (this.opts.showStatus) cli.action.status(this._status('connecting'))
+      this.uri = url.parse(this.dyno.attach_url)
+
+      if (this.uri.protocol === 'http:' || this.uri.protocol === 'https:') {
+        this._ssh()
+      } else {
+        this._rendezvous()
+      }
+    })
+  }
+
+  _rendezvous () {
+    if (this.opts.showStatus) cli.action.status(this._status('starting'))
+    let c = tls.connect(this.uri.port, this.uri.hostname, {rejectUnauthorized: this.heroku.options.rejectUnauthorized})
+    c.setTimeout(1000 * 60 * 20)
+    c.setEncoding('utf8')
+    c.on('connect', () => {
+      c.write(this.uri.path.substr(1) + '\r\n', () => {
+        if (this.opts.showStatus) cli.action.status(this._status('connecting'))
+      })
+    })
+    c.on('data', this._readData(c))
+    c.on('close', () => {
+      this.opts['exit-code'] ? this.reject('No exit code returned') : this.resolve()
+      if (this.unpipeStdin) this.unpipeStdin()
+    })
+    c.on('error', this.reject)
+    process.once('SIGINT', () => c.end())
+  }
+
+  _ssh () {
+    this._wait()
+    this._connect((options) => {
+      const host = options.address
+      const port = options.port
+
+      cli.action.done(this._status(this.dyno.state))
+
+      if (this.opts.listen) {
+        cli.console.log(`listening on port ${options.address}:${options.port} for ssh client`)
+      } else {
+        spawn('ssh', [host, '-p', port, '-oStrictHostKeyChecking=no', '-oUserKnownHostsFile=/dev/null'], {
```

Comment:
the ssh client connects to localhost and a random port on every call of `heroku run` and `heroku run:inside`. your known_hosts will contain countless entries for each port. the ssh client will get slow once you have enough entries. we can't use a single port because that would block the ability to have multiple attached sessions at a time.

i'm more than happy to implement a better solution here, but i don't have one.